pkg/models/vulnerability.go

```diff
@@ -167,3 +174,155 @@ func (v Vulnerability) MarshalYAML() (interface{}, error) {
 
 	return raw, nil
 }
+
+// NormalizedName ensures that the package name is normalized based on ecosystem
+// in accordance to the Vulnerability specification.
+//
+// This is required because currently both GitHub and Pip seem to be a bit
+// inconsistent in their package name handling, so we normalize them
+// to be on the safe side.
+//
+// In the future, it's hoped that this can be improved.
+func (p Package) NormalizedName() string {
+	if p.Ecosystem != lockfile.PipEcosystem {
+		return p.Name
+	}
+
+	// per https://www.python.org/dev/peps/pep-0503/#normalized-names
+	name := regexp.MustCompile(`[-_.]+`).ReplaceAllString(p.Name, "-")
+
+	return strings.ToLower(name)
+}
+
+func (ar Range) containsVersion(pkg lockfile.PackageDetails) bool {
+	if ar.Type != RangeEcosystem && ar.Type != RangeSemVer {
+		return false
+	}
+	// todo: we should probably warn here
+	if len(ar.Events) == 0 {
+		return false
+	}
+
+	vp := semantic.MustParse(pkg.Version, pkg.CompareAs)
+
+	var affected bool
+	for _, e := range ar.Events {
+		if affected {
+			if e.Fixed != "" {
+				affected = vp.CompareStr(e.Fixed) < 0
+			} else if e.LastAffected != "" {
+				affected = e.LastAffected == pkg.Version || vp.CompareStr(e.LastAffected) <= 0
+			}
+		} else if e.Introduced != "" {
+			affected = e.Introduced == "0" || vp.CompareStr(e.Introduced) >= 0
+		}
+	}
```

Comment:
I think this assumes the events are sorted, and would not work correctly if it is not sorted?

The schema says that it's recommended to keep the events sorted, but explicitly says that it is not required: 
https://ossf.github.io/osv-schema/#requirements
> There must be at least one `"introduced"` object in the events array. While not required, it’s also recommended to keep the `events` array sorted according to the `affected[].ranges[].type` of the range.

Reply:
This is an implementation of the algorithm provided [in the spec](https://ossf.github.io/osv-schema/#evaluation) which does assume events are sorted - while it feels like handling sorting could be messy, I'm not sure exactly how complex it would be to support but ultimately I think that would be better supported in a follow up PR (ideally with an update to the spec doc to reflect that part of the algorithm too)